src/sentry/similarity.py

```diff
@@ -272,3 +298,215 @@ def record(self, scope, key, characteristics):
+                key=lambda frame: frame.get('in_app', False),
+            )
+        )
+    )
+
+
+class ExceptionFeature(object):
+    def __init__(self, function):
+        self.function = function
+        self.logger = logging.getLogger(__name__)
+
+    def extract(self, event):
+        try:
+            exceptions = event.data['sentry.interfaces.Exception']['values']
+        except KeyError as error:
+            self.logger.info('Could not extract characteristic(s) from %r due error: %r', event, error, exc_info=True)
+            return
+
+        for exception in exceptions:
+            try:
+                yield self.function(exception)
+            except Exception as error:
+                self.logger.exception('Could not extract characteristic(s) from exception in %r due to error: %r', event, error)
+
+
+class MessageFeature(object):
+    def __init__(self, function):
+        self.function = function
+        self.logger = logging.getLogger(__name__)
+
+    def extract(self, event):
+        try:
+            message = event.data['sentry.interfaces.Message']
+        except KeyError as error:
+            self.logger.info('Could not extract characteristic(s) from %r due error: %r', event, error, exc_info=True)
+            return
+
+        try:
+            yield self.function(message)
+        except Exception as error:
+            self.logger.exception('Could not extract characteristic(s) from message of %r due to error: %r', event, error)
+
+
+class FeatureSet(object):
+    def __init__(self, index, aliases, features):
+        self.index = index
+        self.aliases = aliases
+        self.features = features
+        self.__number_format = get_number_format(0xFFFFFFFF)
+        assert set(self.aliases) == set(self.features)
+
+    def record(self, event):
+        items = []
+        for label, feature in self.features.items():
+            alias = self.aliases[label]
+            scope = ':'.join((
+                alias,
+                self.__number_format.pack(event.project_id),
+            ))
+            for characteristics in feature.extract(event):
+                if characteristics:
+                    items.append((
+                        scope,
+                        self.__number_format.pack(event.group_id),
+                        characteristics,
+                    ))
+        return self.index.record_multi(items)
+
+    def query(self, group):
+        results = {}
+        key = self.__number_format.pack(group.id)
+        for label in self.features.keys():
+            alias = self.aliases[label]
+            scope = ':'.join((
+                alias,
+                self.__number_format.pack(group.project_id)
+            ))
+            results[label] = map(
+                lambda (id, score): (
+                    self.__number_format.unpack(id)[0],
+                    score,
+                ),
+                self.index.query(scope, key)
```

Comment:
Note for the future: need to add a `query_multi` to the index implementation, similar to `record_multi` above.